Defer pg agent listener thread until contexts are initialized

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28030 Minor improvements in RPC api docs
* **#28013 Defer pg agent listener thread until contexts are initialized**

ProcessGroupAgent currently kicks off the listener thread in its
constructor. However, serving requests requires contexts to be
initialized, e.g., RRefContext and agent_ global var in api.py,
which might not be done yet when the first request arrives.
ProcessGroupAgent does not know what would be the appropriate time
to start the listener thread, hence exposing an API for higher
layer code to explicitly start listeners.

Differential Revision: [D17932271](https://our.internmc.facebook.com/intern/diff/D17932271)

mrshenli committed:
Defer pg agent listener thread until contexts are initialized

ProcessGroupAgent currently kicks off the listener thread in its
constructor. However, serving requests requires contexts to be
initialized, e.g., RRefContext and agent_ global var in api.py,
which might not be done yet when the first request arrives.
ProcessGroupAgent does not know what would be the appropriate time
to start the listener thread, hence exposing an API for higher
layer code to explicitly start listeners.

[ghstack-poisoned]

mrshenli commented:
@pytorchbot retest this please

xush6528 commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -103,7 +103,8 @@ PyObject* rpc_init(PyObject* /* unused */) {
 
-  module.def("_init_rpc_agent", [](std::shared_ptr<RpcAgent> agent) {
+  module.def("_start_rpc_agent", [](std::shared_ptr<RpcAgent> agent) {
+    agent->start();
```
Shouldn't we reverse this order?

satgera commented:
LGTM !

mrshenli committed:
Update on "Defer pg agent listener thread until contexts are initialized"


ProcessGroupAgent currently kicks off the listener thread in its
constructor. However, serving requests requires contexts to be
initialized, e.g., RRefContext and agent_ global var in api.py,
which might not be done yet when the first request arrives.
ProcessGroupAgent does not know what would be the appropriate time
to start the listener thread, hence exposing an API for higher
layer code to explicitly start listeners.

Differential Revision: [D17932271](https://our.internmc.facebook.com/intern/diff/D17932271)

[ghstack-poisoned]

xush6528 reviewed (approved):
looks good now!

mrshenli committed:
Update on "Defer pg agent listener thread until contexts are initialized"


ProcessGroupAgent currently kicks off the listener thread in its
constructor. However, serving requests requires contexts to be
initialized, e.g., RRefContext and agent_ global var in api.py,
which might not be done yet when the first request arrives.
ProcessGroupAgent does not know what would be the appropriate time
to start the listener thread, hence exposing an API for higher
layer code to explicitly start listeners.

Differential Revision: [D17932271](https://our.internmc.facebook.com/intern/diff/D17932271)

[ghstack-poisoned]
